app/controllers/work_packages/auto_completes_controller.rb

```diff
@@ -30,43 +30,65 @@
 require 'rack/utils'
 
 class WorkPackages::AutoCompletesController < ApplicationController
-  before_filter :find_project
-
   def index
-    @work_packages = []
-    q = params[:q].to_s
+    scope = determine_scope
+    if scope.nil?
+      render_404
+      return
+    end
 
-    if q.present?
-      query = (params[:scope] == 'all' && Setting.cross_project_work_package_relations?) ? WorkPackage : @project.work_packages
+    query_term = params[:q].to_s
+    @work_packages = []
+    # query for exact ID matches first, to make an exact match the first result of autocompletion
+    if query_term =~ /\A\d+\z/
+      @work_packages |= scope.visible.find_all_by_id(query_term.to_i)
+    end
 
-      @work_packages |= query.visible.find_all_by_id(q.to_i) if q =~ /\A\d+\z/
+    sql_query = ["LOWER(#{WorkPackage.table_name}.subject) LIKE :q OR
+                  CAST(#{WorkPackage.table_name}.id AS CHAR(13)) LIKE :q",
+                 { q: "%#{query_term.downcase}%" }]
 
-      @work_packages |= query.visible.find(:all,
-                                           limit: 10,
-                                           order: "#{WorkPackage.table_name}.id ASC",
-                                           conditions: ["LOWER(#{WorkPackage.table_name}.subject) LIKE :q OR CAST(#{WorkPackage.table_name}.id AS CHAR(13)) LIKE :q", { q: "%#{q.downcase}%" }])
-    end
+    @work_packages |= scope.visible
+                      .where(sql_query)
+                      .order("#{WorkPackage.table_name}.id ASC") # :id does not work because...
+                      .limit(10)
 
     respond_to do |format|
-      format.html { render layout: false }
-      format.any(:xml, :json) { render request.format.to_sym => wp_hash_with_string }
+      format.html do render layout: false end
+      format.any(:xml, :json) { render request.format.to_sym => wp_hashes_with_string }
     end
   end
 
   private
 
-  def wp_hash_with_string
-    @work_packages.map do |wp|
-      Hash[wp.attributes.map do |key, value|
-        [key, Rack::Utils.escape_html(value)]
-      end << ['to_s', Rack::Utils.escape_html(wp.to_s)]]
+  def wp_hashes_with_string
+    @work_packages.map do |work_package|
+      wp_hash = Hash.new
+      work_package.attributes.each do |key, value| wp_hash[key] = Rack::Utils.escape_html(value) end
+      wp_hash['to_s'] = Rack::Utils.escape_html(work_package.to_s)
+      wp_hash
     end
   end
 
   def find_project
     project_id = (params[:work_package] && params[:work_package][:project_id]) || params[:project_id]
-    @project = Project.find(project_id)
+    return nil unless project_id
+    Project.find(project_id)
   rescue ActiveRecord::RecordNotFound
-    render_404
+    nil
+  end
+
+  def determine_scope
+    project = find_project
+
+    if params[:scope] == 'relatable'
+      return nil unless project
+
+      Setting.cross_project_work_package_relations? ? WorkPackage.scoped : project.work_packages
```

Comment:
@NobodysNightmare ohh .. then never mind! I, too, was confused by GitHub showing this in the PR diff . I've actually already dealt with this deprecation in #3194 (see https://github.com/myabc/openproject/commit/0c729dae548a90279c703f820791ef811a43879f#diff-75bc1c1409a39ce317aa768f040e8839L87). Feel free to review :wink: